File: code/2130.maximum-twin-sum-of-a-linked-list.php
<?php
/*
 * @lc app=leetcode id=2130 lang=php
 *
 * [2130] Maximum Twin Sum of a Linked List
 */

// @lc code=start
/**
 * Definition for a singly-linked list.
 * class ListNode {
 *     public $val = 0;
 *     public $next = null;
 *     function __construct($val = 0, $next = null) {
 *         $this->val = $val;
 *         $this->next = $next;
 *     }
 * }
 */
class Solution {

    /**
     * @param ListNode $head
     * @return Integer
     */
    function pairSum($head) {
        $max = 0;

        $slow = $head;
        $fast = $head;
        while ($fast) {
            $fast = $fast->next->next;
            $slow = $slow->next;
        }

        $curr = $slow;
        $prev = null;
        $temp = null;
        while ($curr) {
            $temp = $curr->next;
            $curr->next = $prev;
            $prev = $curr;
            $curr = $temp;
        }

        while ($prev) {
            $sum = $head->val + $prev->val;
            $max = $sum > $max ? $sum : $max;

            $prev = $prev->next;
            $head = $head->next;
        }

        return $max;
    }
}
// @lc code=end

File: code/2130.maximum-twin-sum-of-a-linked-list.go
package code

/*
 * @lc app=leetcode id=2130 lang=golang
 *
 * [2130] Maximum Twin Sum of a Linked List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func pairSum(head *ListNode) int {
	if head.Next.Next == nil {
		return head.Val + head.Next.Val
	}

	// 後半のノードを取得
	slow, fast := head, head.Next.Next
	for fast != nil {
		slow = slow.Next
		fast = fast.Next.Next
	}
	suf := slow.Next

	// 後半のノードを反転
	var prev *ListNode
	current := suf
	for current != nil {
		next := current.Next
		current.Next = prev

		prev = current
		current = next
	}

	result := 0
	for prev != nil {
		result = max(result, head.Val+prev.Val)
		head = head.Next
		prev = prev.Next
	}

	return result
}

// @lc code=end
